test/read.jl

```diff
@@ -0,0 +1,245 @@
+mktempdir() do dir
+
+tasks = []
+
+# Create test file...
+filename = joinpath(dir, "file.txt")
+text = "C1,C2\n1,2\na,b\n"
+
+# List of IO producers...
+l = Vector{Tuple{AbstractString,Function}}()
+
+
+# File
+io = (text) -> begin
+    open(io-> write(io, text), filename, "w")
+    Base.Filesystem.open(filename, Base.Filesystem.JL_O_RDONLY)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, Base.Filesystem.File)
+close(s)
+push!(l, ("File", io))
+
+
+# IOStream
+io = (text) -> begin
+    open(io-> write(io, text), filename, "w")
+    open(filename)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, IOStream)
+close(s)
+push!(l, ("IOStream", io))
+
+
+# IOBuffer
+io = (text)->IOBuffer(text)
+s = io(text)
+@test isa(s, IO)
+@test isa(s, IOBuffer)
+close(s)
+push!(l, ("IOBuffer", io))
+
+
+# TCPSocket
+
+# PR#14627
+Base.connect!(sock::TCPSocket, addr::Base.InetAddr) = Base.connect!(sock, addr.host, addr.port)
+
+addr = Base.InetAddr(ip"127.0.0.1", 4444)
+io = (text) -> begin
+    c = Condition()
+    tsk = @async begin
+        srv = listen(addr)
+        notify(c)
+        sock = accept(srv)
+        write(sock,text)
+        close(sock)
+        close(srv)
+    end
+    push!(tasks, tsk)
+    wait(c)
+    connect(addr)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, TCPSocket)
+close(s)
+push!(l, ("TCPSocket", io))
+
+
+@windows ? nothing : begin
+
+# PipeEndpoint
+socketname = joinpath(dir, "socket")
```

Comment:
I think you're right,
However, I haven't done any serious dev work on Windows for about 15 years, so my memory of it isn't good enough to get it working. I went though a bunch of cycles of push, wait for appveyor, guess, push again... I managed to work around `mktempdir() do ...` being broken on windows, but appveyor is just way too slow for debugging the sockets stuff.

I've [added a note in the test code](https://github.com/samoconnor/julia/commit/0032cd62b397f3917c94380791e0559e73ce8e28#diff-435745fdfa02a0dc4f3b747ba977af11R46) to the effect that this stuff should work on windows but that someone who has the right knowledge and tools will need to look at it.